File: Problems/13-Max-Consecutive-Ones-III/Max-Consecutive-Ones-III.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> prefixSum(n+1, 0);
        for (int i = 1; i <= n; i++) prefixSum[i] = prefixSum[i-1] + nums[i-1];

        int res = 0;
        for (int i = 1; i <= n; i++)
        {
            int left  = i;
            int right = n;

            int ans = 0;
            while (left <= right)
            {
                int mid = (right-left)/2+left;
                int num0 = (mid-i+1)-prefixSum[mid] + prefixSum[i-1];   // in [i..mid]
                if (num0 <= k) {ans = mid; left = mid + 1;}
                else {right = mid-1;}
            }

            res = max(res, ans - i + 1);
        }
        
        return res;
        
    }
};

File: Problems/13-Max-Consecutive-Ones-III/Max_consecutive_ones_III.py
class Solution(object):
    def longestOnes(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        left    ,right = 0,0
        res = 0
        while(right<len(nums)):
            if(nums[right] == 0):
                k-=1
            right+=1
            while(k<0):
                if(nums[left]==0):
                    k+=1
                left+=1
            res = max(res,right-left)
        return res